[Doc Feature][Targeted messages Ph 1.5 - GA][5518210] (#14428)
MicrosoftDocs/msteams-docs: msteams-platform/agents-in-teams/targeted-messages.md

Reviewer: ```suggestion
_Targeted messaging_ enables users and agents to send each other messages in the context of a group conversation that can only be seen by the sender and the targeted recipient.

:::image type="content" source="../assets/images/agents-in-teams/targeted-messages/targeted-messages.png" alt-text="Image shows user scenarios for targeted messages" border="false" lightbox="../assets/images/agents-in-teams/targeted-messages/targeted-messages-main.png":::

With targeted messaging, users and agents can privately interact without distracting other users or exposing information that isn't meant for the group. Consider the following scenarios:

- A user asks an agent to generate a summary of the discussion without the other participants seeing the request or the response.
- A user privately asks for information from an agent with the intent of sharing it with the group, but wants to confirm the contents of the message first. The agent responds privately, with a [suggested action](../bots/how-to/conversations/suggested-actions.md) requesting the user's approval to share its response. When the user approves, the agent resends the message publicly.
- A user directs and monitors background work with agents, referencing the context of the conversation, without distracting the other participants.
- An agent detects a new user entering a large, long-running conversation and proactively send them a private welcome message and a summary.
```

- A user can direct and monitor background work with agents, referencing the context of the conversation, without distracting the other participants.
- An agent can detect a new user entering a large, long-running conversation and proactively send them a private welcome message and a summary.

> [!IMPORTANT]
> Targeted messaging is a fundamental part of agent design for group conversations and requires careful handling of user privacy expectations. For more information, see [best practices and design guidance](#best-practices-and-design-guidance).

### Key aspects of targeted messages

Targeted messages:

- Can be used only for one-to-one interactions between an agent and a user in channels, group chats, and meeting chats.
- Support all the [capabilities of standard messages](../bots/build-conversational-capability.md#message-content) like buttons, images, Adaptive Cards, and files, but don't support reactions, replies, or forwarding.

>     A targeted message can contain normal message content, but Teams doesn't let that targeted message participate in the usual social or threading actions that might expose it more broadly. Targeted messages are private, temporary, and visible to a single user only in a shared conversation, so features that create secondary visible artifacts, shared thread context, or message redistribution are intentionally disabled. The default agent response to a slash command is targeted message, and then on approval to resend it as a public message rather than trying to reply to, react to, or forward the targeted one.

Reviewer: The first part of this is covered in the above bullet; the second part is not correct, "approval to resend it as a public message" is not a default behavior, it requires developer implementation.

Activating an agent's targeted message command switches the compose box to targeted message mode. After the user composes a message and selects **Send**, the resulting message will be targeted to the agent and can't be seen by other participants in the conversation.

For more information about slash commands, including how to register extra named slash commands that can be dispatched to your agent, see [expose slash commands from agents and apps](agent-slash-commands.md).

You can implement the following agent-to-user response flows:

- _Private response mode_ is the default for slash-command responses and keeps the interaction focused between the user and the agent. Use it for drafts, summaries, and personal tasks.
- _Public response_ mode lets the user share the response to the wider audience.

> - _Private-to-public response flow_ enables [user approval for a private response](#approval-workflow-using-suggested-actions) to be shared publicly.

Reviewer: Drop this; this is restating the fundamental ideas of targeted messages that have already been stated, and "modes" sounds like some other first-class feature.

Author: Updated

> ### Prompt Preview in targeted messages

Reviewer: My comment since this article began authoring still stands - unless this distinguishes from quoted replies, this should be dropped.

Author: I initially added the information referenced in the quoted response to the Prompt Preview section, along with a comparison in the Best Practices and Design Guidance document and included a link to it in the Teams SDK article. Following subsequent feedback, I removed the comparison and condensed the content into a shorter paragraph within the Best Practices and Design Guidance document. I have now also added this concise paragraph about quoted replies.

Prompt Preview can appear in both private and public agent responses:

# [Private agent-to-user response](#tab/private)

The agent replies privately to the user's targeted message, so only the intended user can see the reply and the prompt preview.

  :::image type="content" source="../assets/images/agents-in-teams/targeted-messages/private-prompt-preview.png" alt-text="Image shows the prompt preview for private agent-to-user response." lightbox="../assets/images/agents-in-teams/agent-slash-commands/private-prompt-preview.png":::

# [Public agent-to-user response](#tab/public)

The agent sends a public reply to the user's request that includes the prompt preview. It's visible to everyone in the chat.

  :::image type="content" source="../assets/images/agents-in-teams/targeted-messages/public-prompt-preview.png" alt-text="Image shows the prompt preview for public agent-to-user response." lightbox="../assets/images/agents-in-teams/agent-slash-commands/public-prompt-preview.png":::

> [!NOTE]
> When an agent re-shares a private response publicly, the prompt preview shows the original targeted message above the agent response.

---

## Implement targeted messages

> Targeted messages are sent and received using the same operations as [standard single-recipient messages](/microsoftteams/platform/teams-sdk/essentials/sending-messages/overview?pivots=typescript) in the Teams SDK but have a Boolean property indicating whether they're targeted. You can also control response visibility and add prompt preview in the agent responses.

Reviewer: ```suggestion
Targeted messages are sent and received using the same operations as [standard single-recipient messages](/microsoftteams/platform/teams-sdk/essentials/sending-messages/overview?pivots=typescript) in the Teams SDK but have a boolean property indicating whether they're targeted.
```

Response visibility _is_ targeted messaging, it's not something additional to be controlled.

Author: Updated

  :::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/private-prompt-preview.png" alt-text="Image shows the prompt preview for private agent-to-user response." lightbox="../assets/images/agents-in-teams/agent-slash-commands/private-prompt-preview.png":::

# [Public agent-to-user response](#tab/public)

The agent sends a public resply to the user's request that includes the prompt preview. It's visible to everyone in the chat.

  :::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/public-prompt-preview.png" alt-text="Image shows the prompt preview for public agent-to-user response." lightbox="../assets/images/agents-in-teams/agent-slash-commands/public-prompt-preview.png":::

---

Compared with [quoted replies](/microsoftteams/platform/teams-sdk/essentials/sending-messages/overview?pivots=csharp), Prompt Preview is purpose-built for targeted response workflows. Quoted replies point readers back to an earlier message in a thread, while Prompt Preview preserves the private prompt-response context and supports approval-based sharing from private to public.

> ::: zone pivot="typescript"

Reviewer: These pivots are not scoped correctly, they should not include this much content. They should be used in pretty much the same places where tab controls are currently used.

For more about slash commands, including how to register extra named slash commands that can be dispatched to your agent, see [Expose slash commands from agents and apps](agent-slash-commands.md).

> ### Response visibility modes

Reviewer: Remove this section, it sounds like it's trying to introduce a new, more complex concept (_response visibility modes_) on top of the basic concept of message visibility that is already the main point of targeted messaging and is the subject of this whole article, and it mostly contains best practices that should go in the guidance section.

Author: Updated

You can implement the following agent-to-user response flows:

- _Private response mode_ is the default for slash-command responses and keeps the interaction focused between the user and the agent. Use it for drafts, summaries, and personal tasks.
- _Public response_ mode lets the user share the response to the wider audience.
- _Private-to-public response flow_ lets the user approve a private response to be shared publicly.

You can design a user approval workflow for reposting a private targeted message as a public message. User approval matters in targeted messaging because the agent’s first response is intentionally private. For more information, see [best practices and design guidance](#best-practices-and-design-guidance).

> [Suggested actions](../bots/how-to/conversations/suggested-actions.md), such as Approve or Reject, can then provide a lightweight way to confirm sharing decisions while preserving the prompt context. See [best practices and design guidance](#best-practices-and-design-guidance).

Reviewer: Calling this scenario out repeatedly throughout the article is distracting; give it its own H2 above the best practices section and put all the information about it there.

The right docs solution for it long term is to give it its own how-to article that actually shows how to do it in Teams SDK - that not only gets the specific information out there but shines a brighter light on it. @dampbrownshoes

### Prompt preview in targeted messages

> Prompt Preview helps preserve conversational context in targeted messaging scenarios by displaying a compact preview of the user's original request above the agent's response. This is particularly useful when an agent responds to a specific user within a shared conversation. Targeted messages appear inline in the conversation but are visible only to the intended recipient. Prompt Preview helps users understand the context of a response without exposing the original prompt to other participants. A single agent response can contain multiple prompt previews.

Reviewer: ```suggestion
Prompt Preview helps preserve conversational context in targeted messaging scenarios by displaying a compact preview of the user's original request above the agent's response. Prompt Preview helps users understand the context of a response without exposing the original prompt to other participants. A single agent response can contain multiple prompt previews.
```

> Prompt preview can appear in both private and public agent responses. The difference is only who can see the reply after it is sent:

Reviewer: The second sentence and the screenshots don't really make any kind of point about prompt preview, they're just showing private vs. public messages. The interesting aspect here that needs to be covered is "how does a user know if their prompt will be shown when an agent sends a public message, and can they control that and how?" I can imagine many scenarios, especially the important private-to-public scenario, where the user will want to know if their prompt will be visible to others, and ideally have control over that.

Author: Updated

# [Private agent-to-user response](#tab/private)

The agent replies privately to the user's targeted message, so only the intended user can see the reply and the prompt preview.

  :::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/private-prompt-preview.png" alt-text="Image shows the prompt preview for private agent-to-user response." lightbox="../assets/images/agents-in-teams/agent-slash-commands/private-prompt-preview.png":::

# [Public agent-to-user response](#tab/public)

The agent sends a public resply to the user's request that includes the prompt preview. It's visible to everyone in the chat.

  :::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/public-prompt-preview.png" alt-text="Image shows the prompt preview for public agent-to-user response." lightbox="../assets/images/agents-in-teams/agent-slash-commands/public-prompt-preview.png":::

---

> Compared with [quoted replies](/microsoftteams/platform/teams-sdk/essentials/sending-messages/overview?pivots=csharp), Prompt Preview is purpose-built for targeted response workflows. Quoted replies point readers back to an earlier message in a thread, while Prompt Preview preserves the private prompt-response context and supports approval-based sharing from private to public.

Reviewer: This doesn't sufficiently explain how prompt preview is _different_ from quoted replies. In what way does it _>preserve the private prompt-response context and support approval-based sharing_ that quoted replies don't?  What makes the user experience different enough that it's worth explaining to developers, especially given that the developer experience for it is exactly the same and designed to be automatic (use `reply()`)? Ideally we'd have screenshots showing the difference in experience.

Author: Updated in Best practices

Prompt Preview can be enabled when an agent or bot replies to a user’s message through supported messaging APIs. In reactive scenarios, such as replying with `send()` or `reply()` to a targeted message, the SDK can attach the required entity automatically. In proactive scenarios, where the agent or bot sends a message outside the original interaction context, the agent ot bot must attach the entity manually by referencing the targeted message ID. Because Adaptive Cards may not align with this experience in some implementations, suggested actions can be used instead to provide lightweight call-to-action options such as Approve or Reject options while preserving the new prompt preview experience.

### Response visibility modes

Setting `isTargeted: true` with `WithRecipient` delivers this message as a private, user-specific message inside a shared conversation. Setting `isTargeted: false` sets the recipient metadata, but doesn't mark the message as privately targeted. In other words, Teams SDK lets you identify who the logical recipient is without necessarily turning the message into a targeted or ephemeral message.

> This is useful because a message can still be addressed to a specific account for routing or replying while remaining a normal message from a visibility perspective:

Reviewer: This needs to reflect better technical understanding. I asked the eng team about what using `WithRecipient` with `isTargeted:false` does and they said "nothing". What does "addressed to a specific account for routing or replying while remaining a normal message" actually mean, functionally, and how is it useful?

  )
```

::: zone-end

> [!NOTE]
> If attempting to send a targeted message results in an error, consider sending a 1:1 chat message as a fallback.

> ### Integrate prompt preview in a targeted message

Reviewer: This finally starts to hint at a difference between quoted replies and prompt preview - the API surface is different in proactive scenarios. What if, in a proactive scenario, I explicitly apply a quoted reply to a targeted message?

Author: I will need to connect with Nirav for this query. Proactive messages may use the referenced message ID for the quoted response, and the conversation ID. Once Nirav is available, I can confirm the quoted response scenario with him and add it as a fast follow-up.

@nickwalkmsft @dampbrownshoes @niravkapoor

To render prompt preview, include a `targetedMessageInfo` entity in the reply activity and set its `messageId` value to the message ID of the original targeted message to show the user's original request above the agent response, as shown in this example:

```json

"entities": [{
  "type": "targetedMessageInfo",
  "messageId": "xxxxxxxxxxxxx"
}]

```

This example adds a `targetedMessageInfo` entity `type` to the activity’s `entities` array and references the original user message through its `messageId`. This association enables _Prompt Preview_ and enables the original request to appear above the agent response.

### Update or delete a targeted message

Targeted messages can be updated and deleted in the same way as standard messages, with the following limitations:

- The visibility of a targeted message can't be changed.
- Targeted messages expire after 24 hours and are automatically deleted from clients. Attempting to modify or delete an expired message results in an error.

For more information on updating or deleting a targeted message, see [Teams SDK](/microsoftteams/platform/teams-sdk/essentials/sending-messages/overview?pivots=csharp#targeted-messages).

> ### Approval workflow using suggested actions

Reviewer: There was an update planned to the suggested actions article to cover this; what's the status on it?

Author: Suggested actions is being updated in a separate PR, currently under internal review.

User approval is a key safeguard in targeted messaging workflows. Since an agent’s response may contain user-specific information, explicit user consent is required before sharing it in a public conversation.

The agent can respond privately, with a [suggested action](../bots/how-to/conversations/suggested-actions.md) requesting the user's approval to share its response. When the user approves, the agent resends the message publicly. The approval experience can be implemented through interactive review surfaces such as Adaptive Cards or suggested actions. Here's a recommended user approval workflow:

1. Receive the user’s targeted request.
1. Send the agent’s response privately, including a prompt preview for context.
1. Present approval options using Suggested Actions or an Adaptive Card.
1. Publish the response to the conversation only after the user approves sharing.

Recommended actions can include Approve, Reject, Share, and Update.

## Best practices and design guidance

Targeted messaging is a fundamental part of group conversations that include agents. Users expect agents to respect the privacy boundaries it creates unless they approve it for sharing with wider audience.

> Agents that can receive targeted messages should **always** check the visibility of messages received in group contexts and use it to inform response generation and conversation context tracking.

Reviewer: ```suggestion
The ability for users and agents to communicate privately with targeted messaging is a fundamental consideration in group conversations. Agents should **always** consider the visibility of messages they receive and use it to inform response generation, response visibility, and whether the original request should be tracked as part of the conversation's context.
```

Author: Updated

Prompt Preview can be added automatically in _reactive_ scenarios such as when replying with `send()` or `reply()` to a user’s targeted message or slash-command request through the Teams SDK or REST API. These SDK methods typically attach the `targetedMessageInfo` entity automatically, keeping the response associated with the incoming user interaction and tied to the original prompt. In _proactive_ flows, however, the agent sends the response outside the original turn, so it must explicitly add the `targetedMessageInfo` entity and reference the targeted message ID before sending the private or public message, ensuring the preview points to the correct original request.

To render prompt preview, include a `targetedMessageInfo` entity in the reply activity and set its `messageId` value to the message ID of the original targeted message, which Teams uses to show the user's original request above the agent response, as shown in this example:

> ```json

Reviewer 2: Add for Teams sdk as well 

C#:

if (isTargeted)
        {
            response.AddTargetedMessageInfo(targetedMessageId);
        }

TS:

 if (isTargeted) {
            response.addTargetedMessageInfo(targetedMessageId);
        }

Py:

if is_targeted:
            response.add_targeted_message_info(activity.id)

Author: Updated as suggested.

  )
```

::: zone-end

> [!NOTE]
> If attempting to send a targeted message results in an error, consider sending a 1:1 chat message as a fallback.

> Response visibility is controlled by the `WithRecipient` configuration and the `isTargeted` property. Use `isTargeted: true` with `WithRecipient` to deliver a private, user-specific response within a shared conversation. Use `isTargeted: false` with `WithRecipient` only when you need to identify an intended recipient while keeping the message visible according to the normal conversation context. This distinction allows developers to identify an intended recipient without automatically creating a private or ephemeral experience.

Reviewer: Again, this needs confirmation. What does _identifying an intended recipient without automatically creating a private or ephemeral experience_ actually do? My understanding is that this is not actually a feature.

### Integrate prompt preview in a targeted message

Prompt preview can appear in both private and public agent responses, but the implementation mechanism is the same in both cases: the agent includes `targetedMessageInfo` that references the original targeted message. The difference is only who can see the reply after it is sent.

Prompt Preview can be added automatically in _reactive_ scenarios such as when replying with `send()` or `reply()` to a user’s targeted message or slash-command request through the Teams SDK or REST API. These SDK methods typically attach the `targetedMessageInfo` entity automatically, keeping the response associated with the incoming user interaction and tied to the original prompt. In _proactive_ flows, however, the agent sends the response outside the original turn, so it must explicitly add the `targetedMessageInfo` entity and reference the targeted message ID before sending the private or public message, ensuring the preview points to the correct original request.

> To render prompt preview, include a `targetedMessageInfo` entity in the reply activity and set its `messageId` value to the message ID of the original targeted message, which Teams uses to show the user's original request above the agent response, as shown in this example:

Reviewer: This is all too much text and buries the main points. In reactive scenarios, it's automatic with send or reply, in proactive scenarios, use the Teams SDK implementations provided by Ajay. The JSON is an implementation detail that most developers will not need; it can be left in but should not be emphasized.

```json

"entities": [{
  "type": "targetedMessageInfo",
  "messageId": "xxxxxxxxxxxxx"
}]
```

This example demonstrates the `entities` array needed to add prompt preview with `messageId` of the original message on a reply activity with `type` set to `targetedMessageInfo`, so the original message appears in the prompt preview.

> You can follow Teams SDK samples implementing prompt preview. For more information on implementing prompt preview, see [send a targeted message](#send-a-targeted-message) and [Teams SDK](/microsoftteams/platform/teams-sdk/essentials/sending-messages/overview?pivots=csharp).

Reviewer: ```suggestion
```

<br>

> <details>

Reviewer: Remove usage of `<details>`

Author: Updated

<summary><b>Select to view REST API example</b></summary>

> If you are sending replies through REST APIs, use the same `targetedMessageInfo` entity in the activity payload.

Reviewer: If we need to keep this it can be a subheader in the Implement section

Author: Updated

Authorization: Bearer eyJhbGciOiJIUzI1Ni...
Content-Type: application/json
Show more lines
JSON
{
"type": "message",
"from": {
"id": "28:c9e...",
"name": "Contoso"
},
"conversation": {
"id": "x:17I0...",
"name": "Convo1"
},
"recipient": {
"id": "29:1XJ...",
"name": "Megan Bowen"
},
"text": "My bot's reply",
"entities": [
{
"type": "targetedMessageInfo",
"messageId": "1772129782775"
}
]
}
```

Use `isTargetedActivity=true` for the private reply. For a public repost, send the message normally but keep the same `targetedMessageInfo` entity.

</details>

<br>

### Update or delete a targeted message

Targeted messages can be updated and deleted in the same way as standard messages, with the following limitations:

- The visibility of a targeted message can't be changed.
- Targeted messages expire after 24 hours and are automatically deleted from clients. Attempting to modify or delete an expired message results in an error.

For more information on updating or deleting a targeted message, see [Teams SDK](/microsoftteams/platform/teams-sdk/essentials/sending-messages/overview?pivots=csharp).

## Best practices and design guidance

> Targeted messaging is a fundamental part of group conversations that include agents. Users expect agents to respect the privacy boundaries it creates, unless they approve it for sharing with wider audience.

Reviewer: ```suggestion
Targeted messaging is a fundamental part of group conversations that include agents. Users expect agents to respect the privacy boundaries it creates.
```

This is a general statement

Author: Updated as suggested.

- Take care when using Adaptive Cards in targeted messages. Although the message itself is targeted, interactions with a card can still generate public activity that users might not expect.
- User approval is an important safeguard in targeted messaging workflows. Because an agent’s initial response may contain user-specific information, the user should explicitly confirm before the content is published to a shared conversation. Prompt Preview supports this decision by allowing users to review the proposed response in the context of their original request before choosing whether to share it.

  The approval experience can be implemented through interactive review surfaces such as Adaptive Cards or [suggested actions](../bots/how-to/conversations/suggested-actions.md). Recommended actions include Approve, Reject, Share, and Update. Here's a recommended user approval workflow:

  1. Receive the user’s request as a targeted message.
  1. Send the agent's response privately.
  1. Add prompt preview to preserve context.
  1. Use Adaptive Cards or suggested actions for user approval.

>   1. Repost the agent response publicly only after the user approves sharing.

Reviewer: Again, let's remove all of these lengthy callouts through the article, put this scenario in its own H2 and make a couple anchor links to it where appropriate.

Author: Updated

- Use Prompt Preview whenever an agent responds to a targeted user request. Compared with [quoted replies](/microsoftteams/platform/teams-sdk/essentials/sending-messages/overview?pivots=csharp), this helps users understand the relationship between their original prompt and the agent’s response without requiring them to locate the earlier message.

    | Prompt preview | Quoted replies |
    | --- | --- |
    | Preserves private prompt-response context | References an earlier message in a thread |
    | Supports private-to-public sharing workflows | Focuses on thread navigation |

>     | Maintains approval-based sharing patterns | Does not provide targeted-response context |

Reviewer: Remove this; the whole point of prompt preview is that it's automatic in most cases where it's needed. If there's any guidance to provide about prompt preview, it's when and how *not* to use it, and how to grant the user control over whether it's included.
